server/src/main/java/org/elasticsearch/persistent/AllocatedPersistentTask.java

```diff
@@ -99,30 +100,49 @@ boolean markAsCancelled() {
     }
 
     public void markAsCompleted() {
-        completeAndNotifyIfNeeded(null);
+        completeAndNotifyIfNeeded(null, null);
     }
 
     public void markAsFailed(Exception e) {
         if (CancelTasksRequest.DEFAULT_REASON.equals(getReasonCancelled())) {
-            completeAndNotifyIfNeeded(null);
+            completeAndNotifyIfNeeded(null, null);
         } else {
-            completeAndNotifyIfNeeded(e);
+            completeAndNotifyIfNeeded(e, null);
         }
     }
 
-    private void completeAndNotifyIfNeeded(@Nullable Exception failure) {
+    /**
+     * Indicates that this persistent task is no longer going to run on the local node.
+     * This will cause the local task to be terminated, and the associated persistent
+     * task to be reassigned by the master node. The persistent task <em>may</em> be
+     * reassigned to the same node unless separate measures have been taken to prevent
+     * this. The task should complete any graceful shutdown actions before calling this
+     * method.
+     * @param localAbortReason Reason for the task being aborted on this node. This
+     *                         will be recorded as the reason for unassignment of the
+     *                         persistent task.
+     */
+    public void markAsLocallyAborted(String localAbortReason) {
+        completeAndNotifyIfNeeded(null, Objects.requireNonNull(localAbortReason));
+    }
+
+    private void completeAndNotifyIfNeeded(@Nullable Exception failure, @Nullable String localAbortReason) {
+        assert failure == null || localAbortReason == null
+            : "completion notification has both exception " + failure + " and local abort reason " + localAbortReason;
         final State prevState = state.getAndSet(State.COMPLETED);
         if (prevState == State.COMPLETED) {
             logger.warn("attempt to complete task [{}] with id [{}] in the [{}] state", getAction(), getPersistentTaskId(), prevState);
```

Comment:
The logic after this change is non-trivial now, as there are two acceptable starting states and two acceptable finishing states with the other combinations resulting in either an exception or warning.  I left the case of completed -> completed being just a warning, because I don't know if this can occur in production at the moment.  But now completed -> local_abort and local_abort -> completed throw exceptions.